src/com/amazon/corretto/crypto/provider/EvpRsaPrivateKey.java

```diff
@@ -0,0 +1,49 @@
+// Copyright Amazon.com Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package com.amazon.corretto.crypto.provider;
+
+import java.math.BigInteger;
+import java.security.interfaces.RSAPrivateKey;
+
+class EvpRsaPrivateKey extends EvpRsaKey implements RSAPrivateKey {
+    private static final long serialVersionUID = 1;
+    private static native byte[] encodeRsaPrivateKey(long ptr);
+
+    protected BigInteger privateExponent;
+
+    protected static native byte[] getPrivateExponent(long ptr);
+
+    EvpRsaPrivateKey(long ptr) {
+        this(new InternalKey(ptr));
+    }
+
+    EvpRsaPrivateKey(InternalKey key) {
+        super(key, false);
+    }
+
+    @Override
+    public BigInteger getPrivateExponent() {
+        synchronized (this) {
+            if (privateExponent == null) {
+                privateExponent = nativeBN(EvpRsaPrivateKey::getPrivateExponent);
+            }
+        }
+
+        return privateExponent;
+    }
```

Comment:
I notice on methods that return a value, you put a `synchronized (this) {...}` block around the logic instead of marking the method as `synchronized`.  Is there a performance reason for doing it this way?

On a related note -- there are also security-related reasons to not block on `this`, but only block on "private" objects: https://stackoverflow.com/questions/442564/avoid-synchronizedthis-in-java

Reply:
I'm aware of the issues with synchronizing on this and believe that in the majority of cases it isn't a significant issue. I definitely wouldn't describe it as "security related" since everything starts with "Let someone execute arbitrary code in your JVM", at which point you're already in trouble.

This particular method is mis-written because the synchronized block should be *inside* the null check for `privateExponent` and this should be a properly done double-checked lock (which is legal in Java now).

My general rule of thumb is: Synchronize the method if there is no way to avoid the synchronization or only synchronized for part of the method. Otherwise, synchronize using a block for the smallest internal component.

If you do feel strongly about this, I can introduce an object just for locking.

Reply:
> ... this should be a properly done double-checked lock (which is legal in Java now).

I think double-checked locking is fine as long as the `volatile` keyword is used on the field to ensure that its value is not being cached. https://en.wikipedia.org/wiki/Double-checked_locking#Usage_in_Java

Reply:
I will check with Java experts, but the references I'm seeing on that page are about 20 years out of date. My understanding is that you haven't needed `volatile` in a very long time.

Reply:
Turns out I was wrong and volatile is still needed is some cases (so safer to have it in all). This is an excellent (modern) write-up of the issue. https://shipilev.net/blog/2014/safe-public-construction/